File: go800mon/a800mon/monitor/breakpoints.go
package monitor

import (
	"context"
	"fmt"
	"strings"

	. "go800mon/a800mon"
)

type BreakpointsViewer struct {
	BaseWindowComponent
	rpc              *RpcClient
	grid             *GridWidget
	enabled          bool
	clauses          []BreakpointClauseRow
	lastSnapshot     string
	lastStateSeq     uint64
	hasSnapshot      bool
	pendingAdd       [][]BreakpointCondition
	hasPendingAdd    bool
	pendingDelete    *int
	pendingEnabled   *bool
	pendingClear     bool
	refreshRequested bool
	clearDialog      *DialogWidget
	inputWidget      *InputWidget
	inputActive      bool
}

func NewBreakpointsViewer(rpc *RpcClient, window *Window) *BreakpointsViewer {
	grid := NewGridWidget(window)
	grid.SetColumnGap(0)
	grid.AddColumn("index", 0, ColorAddress.Attr(), nil)
	grid.AddColumn("condition", 0, ColorText.Attr(), nil)
	v := &BreakpointsViewer{
		BaseWindowComponent: NewBaseWindowComponent(grid.Window()),
		rpc:                 rpc,
		grid:                grid,
	}
	v.clearDialog = NewDialogWidget(grid.Window())
	v.inputWidget = NewInputWidget(grid.Window())
	v.inputWidget.SetOnChange(v.onInputChange)
	return v
}

func (v *BreakpointsViewer) Update(ctx context.Context) (bool, error) {
	st := State()
	if !st.BreakpointsSupported {
		return false, nil
	}
	changed := false
	if v.pendingClear {
		v.pendingClear = false
		if err := v.rpc.BPClear(ctx); err == nil {
			v.grid.SetSelectedRow(nil)
			v.refreshRequested = true
			changed = true
		}
	}
	if v.pendingDelete != nil {
		idx := *v.pendingDelete
		v.pendingDelete = nil
		if err := v.rpc.BPDeleteClause(ctx, uint16(idx)); err == nil {
			v.refreshRequested = true
			changed = true
		}
	}
	if v.hasPendingAdd {
		clauses := append([][]BreakpointCondition(nil), v.pendingAdd...)
		v.pendingAdd = nil
		v.hasPendingAdd = false
		for _, clause := range clauses {
			conds := append([]BreakpointCondition(nil), clause...)
			if _, err := v.rpc.BPAddClause(ctx, conds); err == nil {
				v.refreshRequested = true
				changed = true
				continue
			}
			break
		}
	}
	if v.pendingEnabled != nil {
		enabled := *v.pendingEnabled
		v.pendingEnabled = nil
		if _, err := v.rpc.BPSetEnabled(ctx, enabled); err == nil {
			v.refreshRequested = true
			changed = true
		}
	}
	if v.hasSnapshot && !v.refreshRequested && v.lastStateSeq == st.StateSeq {
		return changed, nil
	}
	v.lastStateSeq = st.StateSeq

	list, err := v.rpc.BPList(ctx)
	if err != nil {
		return changed, nil
	}
	v.refreshRequested = false
	clauses := make([]BreakpointClauseRow, 0, len(list.Clauses))
	for _, clause := range list.Clauses {
		conds := make([]BreakpointConditionRow, 0, len(clause))
		for _, cond := range clause {
			conds = append(conds, BreakpointConditionRow{
				CondType: cond.Type,
				Op:       cond.Op,
				Addr:     cond.Addr,
				Value:    cond.Value,
			})
		}
		clauses = append(clauses, BreakpointClauseRow{Conditions: conds})
	}
	snapshot := buildBreakpointsSnapshot(list.Enabled, clauses)
	if v.hasSnapshot && snapshot == v.lastSnapshot {
		return changed, nil
	}
	v.hasSnapshot = true
	v.lastSnapshot = snapshot
	v.enabled = list.Enabled
	v.clauses = clauses
	return true, nil
}

func (v *BreakpointsViewer) Render(_force bool) {
	w := v.Window()
	g := v.grid
	ih := w.Height()
	if ih <= 0 {
		return
	}
	overlayRows := 0
	if v.inputActive || (v.clearDialog != nil && v.clearDialog.Active()) {
		overlayRows = 1
	}
	w.SetTagActive("bp_enabled", v.enabled)
	g.SetViewport(overlayRows, max(0, ih-overlayRows))
	rows := make([][]string, 0, len(v.clauses)+1)

	if len(v.clauses) == 0 {
		rows = append(rows, []string{"", "No breakpoint clauses."})
		g.SetSelectedRow(nil)
	} else {
		for i, clause := range v.clauses {
			rows = append(rows, []string{
				fmt.Sprintf("#%02d ", i+1),
				v.formatClauseText(clause),
			})
		}
	}
	g.SetData(rows)
	g.Render()

	if v.clearDialog != nil && v.clearDialog.Active() {
		v.clearDialog.Render()
	} else if v.inputActive {
		v.inputWidget.Render(false)
	}
}

func (v *BreakpointsViewer) HandleInput(ch int) bool {
	if v.clearDialog != nil && v.clearDialog.Active() {
		result := v.clearDialog.HandleInput(ch)
		if result == DialogInputConfirm {
			v.pendingClear = true
		}
		return !(result == DialogInputNone)
	}
	if ch == '/' {
		v.openInput("")
		return true
	}
	if v.grid.HandleInput(ch) {
		return true
	}
	if ch == KeyDelete() || ch == 330 {
		v.queueDeleteSelected()
		return true
	}
	if ch == 'c' || ch == 'C' {
		v.clearDialog.Activate("Clear all breakpoints?", "YES")
		return true
	}
	if ch == ' ' || ch == 'e' || ch == 'E' {
		enabled := !v.enabled
		v.pendingEnabled = &enabled
		return true
	}
	return false
}

func (v *BreakpointsViewer) formatClauseText(clause BreakpointClauseRow) string {
	parts := make([]string, 0, len(clause.Conditions))
	for i, cond := range clause.Conditions {
		_ = i
		parts = append(parts, v.formatConditionText(cond))
	}
	return strings.Join(parts, " AND ")
}

func (v *BreakpointsViewer) formatConditionText(cond BreakpointConditionRow) string {
	op := BPOpSymbol(cond.Op)
	if cond.CondType == 9 {
		if cond.CondType == 2 || cond.CondType == 3 || cond.CondType == 4 || cond.CondType == 5 {
			return fmt.Sprintf("mem[%s] %s %02X", formatHex16(cond.Addr), op, cond.Value)
		}
		return fmt.Sprintf("mem[%s] %s %s", formatHex16(cond.Addr), op, formatHex16(cond.Value))
	}
	left := BPTypeName(cond.CondType)
	if cond.CondType == 2 || cond.CondType == 3 || cond.CondType == 4 || cond.CondType == 5 {
		return fmt.Sprintf("%s %s %02X", left, op, cond.Value)
	}
	return fmt.Sprintf("%s %s %s", left, op, formatHex16(cond.Value))
}

func (v *BreakpointsViewer) queueDeleteSelected() {
	idx, ok := v.grid.SelectedRow()
	if !ok {
		return
	}
	v.pendingDelete = &idx
	if idx >= len(v.clauses)-1 {
		if len(v.clauses) <= 1 {
			v.grid.SetSelectedRow(nil)
		} else {
			next := len(v.clauses) - 2
			v.grid.SetSelectedRow(&next)
		}
	}
}

func (v *BreakpointsViewer) openInput(initial string) {
	v.inputActive = true
	v.inputWidget.Activate(initial)
	v.inputWidget.SetInvalid(false)
	if app := v.App(); app != nil {
		app.DispatchAction(ActionSetInputFocus, v.handleTextInput)
	}
}

func (v *BreakpointsViewer) closeInput() {
	v.inputActive = false
	v.inputWidget.SetInvalid(false)
	v.inputWidget.Deactivate()
	if app := v.App(); app != nil {
		app.DispatchAction(ActionSetInputFocus, nil)
	}
}

func (v *BreakpointsViewer) onInputChange(text string) {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		v.inputWidget.SetInvalid(false)
		return
	}
	if _, err := ParseBPClauses(trimmed); err != nil {
		v.inputWidget.SetInvalid(true)
		return
	}
	v.inputWidget.SetInvalid(false)
}

func (v *BreakpointsViewer) handleTextInput(ch int) bool {
	if ch == 27 {
		v.closeInput()
		return true
	}
	if ch == 10 || ch == 13 || ch == KeyEnter() {
		if v.inputWidget.Invalid() {
			return true
		}
		trimmed := strings.TrimSpace(v.inputWidget.Buffer())
		if trimmed != "" {
			clauses, err := ParseBPClauses(trimmed)
			if err != nil {
				v.inputWidget.SetInvalid(true)
				return true
			}
			v.pendingAdd = append([][]BreakpointCondition(nil), clauses...)
			v.hasPendingAdd = true
		}
		v.closeInput()
		return true
	}
	if v.inputWidget.HandleKey(ch) {
		return true
	}
	return true
}

func buildBreakpointsSnapshot(enabled bool, clauses []BreakpointClauseRow) string {
	parts := make([]string, 0, len(clauses)+1)
	parts = append(parts, fmt.Sprintf("enabled:%t", enabled))
	for _, clause := range clauses {
		items := make([]string, 0, len(clause.Conditions))
		for _, cond := range clause.Conditions {
			items = append(items, fmt.Sprintf("%d:%d:%04X:%04X", cond.CondType, cond.Op, cond.Addr, cond.Value))
		}
		parts = append(parts, strings.Join(items, ","))
	}
	return strings.Join(parts, "|")
}

File: py800mon/monitor/breakpoints.py
import curses

from ..actions import Actions
from ..app import VisualRpcComponent
from ..breakpoints import format_bp_condition, parse_bp_clauses
from ..datastructures import BreakpointClauseEntry
from ..rpc import RpcException
from ..ui import Color, DialogInput, DialogWidget, GridWidget
from ..ui import InputWidget
from .appstate import state


class BreakpointsViewer(VisualRpcComponent):
    def __init__(self, rpc, window):
        super().__init__(rpc, window)
        self.grid = GridWidget(window, col_gap=0)
        self.grid.add_column("index", width=0, attr=Color.ADDRESS.attr())
        self.grid.add_column("expr", width=0, attr=Color.TEXT.attr())
        self._last_snapshot = None
        self._last_state_seq = None
        self._enabled = False
        self._rows = []
        self._input_active = False
        self._pending_add_clauses = None
        self._pending_delete = None
        self._pending_clear = False
        self._pending_enabled = None
        self._refresh_requested = False
        self._clear_dialog = DialogWidget(self.window)
        self._input_widget = InputWidget(
            self.window,
            max_length=None,
            on_change=self._on_input_change,
        )

    async def update(self):
        if not state.breakpoints_supported:
            return False
        changed = await self._apply_pending_ops()
        cur_seq = state.state_seq
        if (
            not self._refresh_requested
            and self._last_state_seq == cur_seq
            and self._last_snapshot
        ):
            return changed
        self._last_state_seq = cur_seq
        try:
            bp = await self.rpc.breakpoint_list()
        except RpcException:
            return changed
        self._refresh_requested = False
        snapshot = bp
        if snapshot == self._last_snapshot:
            return changed
        self._last_snapshot = snapshot
        self._enabled = bp.enabled
        self._rows = list(bp.clauses)
        return True

    def render(self, force_redraw=False):
        ih = self.window._ih
        if ih <= 0:
            return
        overlay_rows = 1 if (self._input_active or self._clear_dialog.active) else 0
        self.window.set_tag_active("bp_enabled", self._enabled)
        self.grid.set_viewport(y=overlay_rows, height=max(0, ih - overlay_rows))

        rows = []
        if not self._rows:
            rows.append(("", "No breakpoint clauses."))
            self.grid.set_selected_row(None)
        else:
            for idx, clause in enumerate(self._rows, start=1):
                rows.append((f"#{idx:02d} ", self._format_clause_text(clause)))
            selected = self.grid.selected_row
            if selected is None:
                self.grid.set_selected_row(None)
            else:
                self.grid.set_selected_row(selected)
        self.grid.set_data(rows)
        self.grid.render()

        if self._clear_dialog.active:
            self._clear_dialog.render()
        elif self._input_active:
            self.window.cursor = (0, 0)
            color = Color.INPUT_INVALID if self._input_widget.invalid else Color.TEXT
            attr = color.attr() | curses.A_REVERSE
            self.window.print(self._input_widget.buffer, attr=attr)
            self.window.fill_to_eol(attr=attr)

    def _format_clause_text(self, clause):
        if not isinstance(clause, BreakpointClauseEntry):
            return str(clause)
        parts = []
        for cond in clause.conditions:
            parts.append(format_bp_condition(cond))
        return " AND ".join(parts)

    def handle_input(self, ch):
        if self._clear_dialog.active:
            result = self._clear_dialog.handle_input(ch)
            if result == DialogInput.CONFIRM:
                self._pending_clear = True
            return not result == DialogInput.NONE
        if ch == ord("/"):
            self._open_input("")
            return True
        if self.grid.handle_input(ch):
            return True
        if ch in (curses.KEY_DC, 330):
            self._queue_delete_selected()
            return True
        if ch in (ord("c"), ord("C")):
            self._clear_dialog.activate("Clear all breakpoints?", "YES")
            return True
        if ch == ord(" ") or ch in (ord("e"), ord("E")):
            self._pending_enabled = not self._enabled
            return True
        return False

    def _queue_delete_selected(self):
        idx = self.grid.selected_row
        if idx is None:
            return
        self._pending_delete = idx
        if idx >= len(self._rows) - 1:
            if len(self._rows) <= 1:
                self.grid.set_selected_row(None)
            else:
                self.grid.set_selected_row(len(self._rows) - 2)

    def _close_input(self):
        self._input_active = False
        self._input_widget.set_invalid(False)
        self._input_widget.deactivate()
        self.app.dispatch_action(Actions.SET_INPUT_FOCUS, None)
        try:
            curses.curs_set(0)
        except curses.error:
            pass

    def _open_input(self, initial: str):
        self._input_active = True
        self._input_widget.activate(str(initial))
        self._input_widget.set_invalid(False)
        self.app.dispatch_action(Actions.SET_INPUT_FOCUS, self._handle_text_input)
        try:
            curses.curs_set(1)
        except curses.error:
            pass

    def _on_input_change(self, query):
        text = str(query).strip()
        if not text:
            self._input_widget.set_invalid(False)
            return
        try:
            parse_bp_clauses(text)
            self._input_widget.set_invalid(False)
        except ValueError:
            self._input_widget.set_invalid(True)

    def _handle_text_input(self, ch):
        if ch == 27:
            self._close_input()
            return True
        if ch in (10, 13, curses.KEY_ENTER):
            if self._input_widget.invalid:
                return True
            text = self._input_widget.buffer.strip()
            if text:
                try:
                    self._pending_add_clauses = parse_bp_clauses(text)
                except ValueError:
                    self._input_widget.set_invalid(True)
                    return True
            self._close_input()
            return True
        if self._input_widget.handle_key(ch):
            return True
        return True

    async def _apply_pending_ops(self):
        changed = False
        if self._pending_clear:
            self._pending_clear = False
            try:
                await self.rpc.breakpoint_clear()
                self.grid.set_selected_row(None)
                self._refresh_requested = True
                changed = True
            except RpcException:
                pass
        if self._pending_delete is not None:
            idx = self._pending_delete
            self._pending_delete = None
            try:
                await self.rpc.breakpoint_delete_clause(idx)
                self._refresh_requested = True
                changed = True
            except RpcException:
                pass
        if self._pending_enabled is not None:
            enabled = self._pending_enabled
            self._pending_enabled = None
            try:
                await self.rpc.breakpoint_set_enabled(enabled)
                self._refresh_requested = True
                changed = True
            except RpcException:
                pass
        if self._pending_add_clauses is not None:
            clauses = list(self._pending_add_clauses)
            self._pending_add_clauses = None
            for clause in clauses:
                try:
                    await self.rpc.breakpoint_add_clause(list(clause))
                    self._refresh_requested = True
                    changed = True
                except RpcException:
                    break
        return changed
